src/nft.sol

```diff
@@ -13,185 +13,156 @@
-       if (payload == 0)
-         return "0";
-       // calculate string length
-       uint i = payload;
-       uint length;
-
-       while (i != 0) {
-         length++;
-         i = i >> 4;
-       }
-       // parse byte by byte and construct the string
-       i = payload;
-       uint mask = 15;
-       bytes memory result = new bytes(length);
-       uint k = length - 1;
-
-       while (i != 0) {
-         uint curr = (i & mask);
-         result[k--] = curr > 9 ? byte(55 + uint8(curr)) : byte(48 + uint8(curr));
-         i = i >> 4;
-       }
-
-       return string(result);
-     }
-
-    function equalBytes(bytes memory a, bytes memory b) internal pure returns (bool) {
-        if (a.length == b.length) {
-            for (uint i = 0; i < a.length; i++) {
-                if (a[i] != b[i]) {
-                    return false;
-                }
-            }
-            return true;
+    /**
+     * Computes the asset hash from the given params and verifies if its present in the AssetManager
+     * @param to address deposit address of the token
+     * @param properties bytes[] properties of the each proof
+     * @param values bytes[] value associated with each property
+     * @param salts bytes[] salt associated with each property
+     */
+    function _verifyAsset(address to, bytes[] memory properties, bytes[] memory values, bytes32[] memory salts) internal view returns (bool) {
+        require(to != address(0), "not a valid address");
+
+        // construct assetHash from the props, values, salts
+        // append to address
+        bytes memory data = abi.encodePacked(to);
+
+        // append hashes
+        for (uint i=0; i< properties.length; i++) {
+            data = abi.encodePacked(data, keccak256(abi.encodePacked(properties[i], values[i], salts[i])));
         }
-        return false;
+
+        return assetManager.isAssetValid(keccak256(data));
     }
 
-    // --- NFT ---
-    function _checkAnchor(uint anchor, bytes32 data_root, bytes32 sig_root) internal view returns (bool) {
-        bytes32 doc_root;
-        (, doc_root, ) = anchors.getAnchorById(anchor);
-        if (data_root < sig_root) {
-           return doc_root == sha256(concat(data_root, sig_root));
-        } else {
-           return doc_root == sha256(concat(sig_root, data_root));
-        }
+    /**
+     * @dev Checks that provided document is signed by the given identity. Validates and checks if the public key used is a valid SIGNING_KEY.
+     * Does not check if the signature root is part of the document root.
+     * @param data_root bytes32 hash of all data fields of the document which are signed
+     * @param signature bytes The signature used to sign the data root
+     */
+    function _signed(bytes32 data_root, bytes memory signature) internal view {
+        // Extract the public key and identity address from the signature
+        address identity_ = data_root.toEthSignedMessageHash().recover(signature);
```

Comment:
```suggestion
        address key_ = data_root.toEthSignedMessageHash().recover(signature);
```

Reply:
Since we append the transition flag to the data_root/signing_root then this recover will not work, it will never recover the same key.
Some time back I created a custom `toEthSignedMessageHash().recover()` function that needs to be used/moved here. Here is the code snippet: 
https://github.com/centrifuge/centrifuge-ethereum-contracts/commit/05ea88324367e29eb2b5bd872b2d9819cd3c9cbc#diff-52561fd7440284be9f029c743cdcf837R111-R133